common/nexus/payload_serializer_test.go

```diff
@@ -0,0 +1,231 @@
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package nexus
+
+import (
+	"testing"
+
+	"github.com/nexus-rpc/sdk-go/nexus"
+	"github.com/stretchr/testify/require"
+	commonpb "go.temporal.io/api/common/v1"
+	"go.temporal.io/sdk/converter"
+)
+
+func mustToPayload(v any) *commonpb.Payload {
+	conv := converter.GetDefaultDataConverter()
+	payload, err := conv.ToPayload(v)
+	if err != nil {
+		panic(err)
+	}
+	return payload
+}
+
+func TestNexusPayloadSerializer(t *testing.T) {
+	t.Parallel()
+
+	jsonProtoPayload := mustToPayload(commonpb.RetryPolicy{})
+	delete(jsonProtoPayload.Metadata, "messageType")
+	type testcase struct {
+		name         string
+		inputPayload *commonpb.Payload
+		// defaults to inputPayload
+		expectedPayload *commonpb.Payload
+		header          nexus.Header
+	}
+	cases := []testcase{
+		{
+			name:         "json",
+			inputPayload: mustToPayload("foo"),
+			header:       nexus.Header{"type": "application/json"},
+		},
+		{
+			name:         "bytes",
+			inputPayload: mustToPayload([]byte("foo")),
+			header:       nexus.Header{"type": "application/octet-stream"},
+		},
+		{
+			name:         "nil",
+			inputPayload: mustToPayload(nil),
+			header:       nexus.Header{},
+		},
+		{
+			name:         "json proto",
+			inputPayload: jsonProtoPayload,
+			header: nexus.Header{
+				"type": "application/json; format=protobuf",
+			},
+		},
+		{
+			name:         "json proto",
+			inputPayload: mustToPayload(commonpb.RetryPolicy{}),
+			header: nexus.Header{
+				"type": `application/json; format=protobuf; message-type="temporal.api.common.v1.RetryPolicy"`,
+			},
+		},
+		{
+			name: "binary proto",
+			inputPayload: &commonpb.Payload{
+				Data: []byte("ignored"),
+				Metadata: map[string][]byte{
+					"encoding": []byte("binary/protobuf"),
+				},
+			},
+			header: nexus.Header{
+				"type": "application/x-protobuf",
+			},
+		},
+		{
+			name: "binary proto with messageType",
+			inputPayload: &commonpb.Payload{
+				Data: []byte("ignored"),
+				Metadata: map[string][]byte{
+					"encoding":    []byte("binary/protobuf"),
+					"messageType": []byte("temporal.api.common.v1.RetryPolicy"),
+				},
+			},
+			header: nexus.Header{
+				"type": `application/x-protobuf; message-type="temporal.api.common.v1.RetryPolicy"`,
+			},
+		},
+		{
+			name:         "nil pointer",
+			inputPayload: nil,
+			expectedPayload: &commonpb.Payload{
+				Metadata: map[string][]byte{
+					"encoding": []byte("binary/null"),
+				},
+			},
+			header: nexus.Header{},
```

Comment:
Nit: This is the zero value and needn't be specified

Reply:
`nexus.Header{}` is a `map[string]string`, the zero value for a map is nil.